airbyte-cdk/python/unit_tests/sources/test_connector_state_manager.py

```diff
@@ -25,15 +303,116 @@
             },
             {"actors": {"created_at": "1962-10-22"}, "actresses": {"id": "seehorn_rhea"}},
             does_not_raise(),
-            id="test_supports_legacy_json_blob",
+            id="test_get_legacy_from_legacy_state_blob",
+        ),
+        pytest.param(
+            [
+                AirbyteStateMessage(
+                    type=AirbyteStateType.STREAM,
+                    stream=AirbyteStreamState(
+                        stream_descriptor=StreamDescriptor(name="actresses", namespace="public"),
+                        stream_state=None,
+                    ),
+                )
+            ],
+            {"actresses": {}},
+            does_not_raise(),
+            id="test_get_legacy_from_stream_state",
         ),
-        pytest.param({}, {}, does_not_raise(), id="test_initialize_empty_mapping_by_default"),
-        pytest.param([], {}, does_not_raise(), id="test_initialize_empty_state"),
-        pytest.param("strings_are_not_allowed", None, pytest.raises(ValueError), id="test_value_error_is_raised_on_invalid_state_input"),
     ],
 )
 def test_get_legacy_state(input_state, expected_legacy_state, expected_error):
     with expected_error:
-        state_manager = ConnectorStateManager(input_state)
+        state_manager = ConnectorStateManager({}, input_state)
         actual_legacy_state = state_manager.get_legacy_state()
         assert actual_legacy_state == expected_legacy_state
+
+
+def test_get_state_returns_deep_copy():
+    input_state = [
+        AirbyteStateMessage(
+            type=AirbyteStateType.STREAM,
+            stream=AirbyteStreamState(
+                stream_descriptor=StreamDescriptor(name="episodes", namespace="public"),
+                stream_state=AirbyteStateBlob.parse_obj({"id": [109]}),
+            ),
+        )
+    ]
+    state_manager = ConnectorStateManager({}, input_state)
+
+    per_stream_state = state_manager.get_stream_state("episodes", "public")
+    per_stream_state["id"].append(309)
+
+    legacy_state = state_manager.get_legacy_state()
+    legacy_state["episodes"]["id"].append(309)
+
+    assert state_manager.get_stream_state("episodes", "public") == {"id": [109]}
+    assert state_manager.get_legacy_state() == {"episodes": {"id": [109]}}
+
+
+@pytest.mark.parametrize(
+    "start_state, update_name, update_namespace, update_value, expected_legacy_state",
+    [
+        pytest.param(
+            [
+                {
+                    "type": AirbyteStateType.STREAM,
+                    "stream": {"stream_descriptor": {"name": "actors", "namespace": "public"}, "stream_state": {"id": "mckean_michael"}},
```

Comment:
these tests should make sure no chicanery gets past

Reply:
That test! Are you telling me that a test just happens to fail like that? No!

Reply:
Ahah, Google helped me understand this is a Better Call Saul reference.